src/ray/core_worker/pubsub/pubsub_coordinator.cc

```diff
@@ -0,0 +1,199 @@
+  // If code reaches this line, that means the subscriber id was in the index.
+  RAY_CHECK(subscriber_it != subscribers.end());
+  subscribers.erase(subscriber_it);
+  if (subscribers.size() == 0) {
+    objects_to_subscribers_.erase(objects_to_subscribers_it);
+  }
+  return 1;
+}
+
+bool Subscriber::Connect(LongPollConnectCallback long_polling_reply_callback) {
+  if (long_polling_reply_callback_ == nullptr) {
+    long_polling_reply_callback_ = long_polling_reply_callback;
+    return true;
+  }
+  return false;
+}
+
+void Subscriber::QueueMessage(const ObjectID &object_id, bool try_publish) {
+  mailbox_.push_back(object_id);
+  if (try_publish) {
+    PublishIfPossible();
+  }
+}
+
+bool Subscriber::PublishIfPossible(bool force) {
+  if (long_polling_reply_callback_ == nullptr) {
+    return false;
+  }
+  if (force || mailbox_.size() > 0) {
+    long_polling_reply_callback_(mailbox_);
+    long_polling_reply_callback_ = nullptr;
+    mailbox_.clear();
+    return true;
+  }
+  return false;
+}
+
+void PubsubCoordinator::Connect(const NodeID &subscriber_node_id,
+                                LongPollConnectCallback long_poll_connect_callback) {
+  absl::MutexLock lock(&mutex_);
+  RAY_LOG(DEBUG) << "Long polling connection is initiated by " << subscriber_node_id;
+  RAY_CHECK(long_poll_connect_callback != nullptr);
+
+  if (is_node_dead_(subscriber_node_id)) {
+    std::vector<ObjectID> empty_ids;
+    // Reply to the request so that it is properly GC'ed.
+    long_poll_connect_callback(empty_ids);
+    return;
+  }
+
+  if (subscribers_.count(subscriber_node_id) == 0) {
+    subscribers_.emplace(subscriber_node_id, std::make_shared<Subscriber>());
+  }
+  auto subscriber = subscribers_[subscriber_node_id];
+
+  // Since the long polling connection is synchronous between the client and coordinator,
+  // when it connects, the connection shouldn't have existed.
+  RAY_CHECK(subscriber->Connect(std::move(long_poll_connect_callback)));
+  subscriber->PublishIfPossible();
+}
+
+void PubsubCoordinator::RegisterSubscription(const NodeID &subscriber_node_id,
+                                             const ObjectID &object_id) {
+  absl::MutexLock lock(&mutex_);
+  RAY_LOG(DEBUG) << "object id " << object_id << " is subscribed by "
+                 << subscriber_node_id;
+  if (is_node_dead_(subscriber_node_id)) {
+    return;
+  }
+
+  if (subscribers_.count(subscriber_node_id) == 0) {
+    subscribers_.emplace(subscriber_node_id, std::make_shared<Subscriber>());
+  }
+  subscription_index_.AddEntry(object_id, subscriber_node_id);
+}
+
+void PubsubCoordinator::Publish(const ObjectID &object_id) {
+  absl::MutexLock lock(&mutex_);
+  for (const auto &subscriber_id :
+       subscription_index_.GetSubscriberIdsByObjectId(object_id)) {
+    auto it = subscribers_.find(subscriber_id);
+    RAY_CHECK(it != subscribers_.end());
+    auto subscriber = it->second;
+    subscriber->QueueMessage(object_id);
```

Comment:
Supernit: I think that this might avoid an unnecessary shared pointer copy and reference count? Not sure if it's worth the readability hit.
```suggestion
    it->second->QueueMessage(object_id);
```

Reply:
I think it is okay to keep it for readability. I will use the reference if you are concerend about this. 

```
auto &subscriber  = it->second;
```

Reply:
Obviously be careful when bypassing the shared pointer reference counting like this, since (I think?) the reference count could go to 0 and the subscriber could be deallocated while you're still trying to use the reference.